fl4health/clients/nnunet_client.py

```diff
@@ -0,0 +1,795 @@
+            if len(target) > 1:
+                # If deep supervision is in use, we drop the additional targets
+                # when calculating the metrics as we only care about the
+                # original target which by default in nnunet is at index 0
+                m_target = convert_deepsupervision_dict_to_list(target)[0]
+            else:
+                m_target = list(target.values())[0]
+        else:
+            raise TypeError("Was expecting target to be type Dict[str, torch.Tensor] or torch.Tensor")
+
+        # Check if target is one hot encoded. Prediction always is for nnunet
+        # Add channel dimension if there isn't one
+        if m_pred.ndim != m_target.ndim:
+            m_target = m_target.view(m_target.shape[0], 1, *m_target.shape[1:])
+
+        # One hot encode targets if needed
+        if m_pred.shape != m_target.shape:
+            m_target_one_hot = torch.zeros(m_pred.shape, device=self.device, dtype=torch.bool)
+            # This is how nnunet does ohe in their functions
+            # Its a weird function that is not intuitive
+            # CAREFUL: Notice the underscore at the end of the scatter function.
+            # It makes a difference, was a hard bug to find!
+            m_target_one_hot.scatter_(1, m_target.long(), 1)
+        else:
+            m_target_one_hot = m_target
+
+        # Check if ignore label is in use. The nnunet loss figures this out on
+        # it's own, but we do it it manually here for the metrics
+        if self.nnunet_trainer.label_manager.ignore_label is not None:
+            m_pred, m_target_one_hot = self.mask_data(m_pred, m_target_one_hot)
+
+        # m_pred is one hot encoded output logits. Maybe masked by ignore label
+        # m_target_one_hot is one hot encoded boolean label. Maybe masked by ignore label
+        metric_manager.update({"prediction": m_pred}, m_target_one_hot)
+
+    def empty_cache(self) -> None:
+        """
+        Checks torch device and empties cache before training to optimize VRAM usage
+        """
+        if self.device.type == "cuda":
+            torch.cuda.empty_cache()
+        elif self.device.type == "mps":
+            torch.mps.empty_cache()
+        else:
+            pass
+
+    def update_after_step(self, step: int, current_round: Optional[int] = None) -> None:
+        """
+        Update the learning rate. LR Scheduler is initialized in self.get_lr_scheduler.
+
+        Args:
+            step (int): The current step within the context of the entire round
+            current_round (Optional[int], optional). The current FL round. If left as
+                None, we assume the model is finetuning and do not change the learning
+                rate from what it was at the end of training
+        """
+        if current_round is not None:
+            # Determine total number of steps that have occurred so far
+            current_overall_step = ((current_round - 1) * self.steps_per_round) + step
```

Comment:
Here you can use self.total_steps from BasicClient how you are using current_overall_step. self.total_steps will be more accurate as well because this current_overall_step assumes that we are sampling all clients every round